### PR TITLE
Align columns that are offset by single/multiple tabs

### LINKED ISSUE #2943: Inconsistent stripping of whitespace > Offset columns
I apologize if this is user error, but I think this may be a problem with how whitespace is stripped from the first line versus underlying lines

- How to reproduce the bug?
  I have a data file which has some junk at the top and then the lines with data each start with a  tab (\t). That is, the whole table is moved over by a single tab.
  Sample file is attached, using generic mock data. I set the `skip_to_line` argument well below the header and `header=False`
  Commands
`S = dt.fread(file=src_data, sep='\t', skip_to_line=6, header=False, skip_blank_lines=True, strip_whitespace=True, fill=True, verbose=True)`
(fill=False results in an error)
 
- What was the expected behavior?
  Data values aligned under column headings
- What happens?
 First row is aligned, with following rows are offset to the right by a column.
`S.head()`
```
  | C0          C1          C2          C3             C4      C5 
-- + ----------  ----------  ----------  -------------  ------  ---
 0 | 01/25/2021  2.0         Mary Brown  180.00         USD     NA
 1 |             01/25/2021  1.0         Robert Taylor  150.00  USD
 2 |             01/25/2021  1.0         James Smith    35.00   USD
```

- Your environment?
  Datatable version 0.11.1,  Python 3.8.6 [MSC v.1927 64 bit (AMD64)], and OS WIndows 10 version 1909

### PR DESCRIPTION
**Pull request by pradkrish**
Closes #2943 

This is a partial solution that fixes misaligned columns. It doesn't solve creation of an unnamed columns of NA.

### CODE REVIEW

File: src/core/csv/reader.cc [hunk 1]
```diff
@@ -789,8 +789,8 @@ void GenericReader::detect_and_skip_bom() {
 void GenericReader::skip_initial_whitespace() {
   const char* ch = sof;
   if (!sof) return;
-  while ((ch < eof) && (*ch <= ' ') &&
-         (*ch==' ' || *ch=='\n' || *ch=='\r' || *ch=='\t')) {
+  while ((ch < eof) && (*ch <= ' ')  &&
+         (*ch==' ' || *ch=='\n' || *ch=='\r' || (*ch=='\t' && *(ch+1) <= ' '))) {
```
Comment (st-pasha) [inline src/core/csv/reader.cc:793]: Need to also check that `(ch+1) < eof`, otherwise, you might be dereferencing an invalid pointer.

**Review by st-pasha — COMMENTED** [review]
could you also add a test, so that the problem won't reappear in the future?

File: src/core/csv/reader.cc (hunk 1 above)
Comment (pradkrish) [reply src/core/csv/reader.cc]: Done

**Comment by st-pasha** [comment]
One thing I don't like about the tab character `\t` is that it is both a whitespace character, and a legitimate separator. In fact, it is the second most popular separator in use after a comma.
Thus, the definition of a "whitespace" changes when `\t` is a separator -- we don't want it to be treated as the separator in one part of a string, but as whitespace in another.

For example, the following is legitimate CSV fragment:
```
A\tB\tC\n
\t2\t3\n
4\t5\t6\n
```
Note how `\t` at the beginning of a second line is not considered "whitespace at the start of a line".

File: src/core/csv/reader.cc [hunk 2]
```diff
@@ -789,8 +789,9 @@ void GenericReader::detect_and_skip_bom() {
 void GenericReader::skip_initial_whitespace() {
   const char* ch = sof;
   if (!sof) return;
-  while ((ch < eof) && (*ch <= ' ') &&
-         (*ch==' ' || *ch=='\n' || *ch=='\r' || *ch=='\t')) {
+  while ((ch < eof) && (*ch <= ' ')  &&
+         (*ch==' ' || *ch=='\n' || *ch=='\r' ||
+         (*ch=='\t' && (ch+1 < eof) && *(ch+1) != '\t' && *(ch+1) <= ' '))) {
```
[the remark this answers is not in the document]
Comment (pradkrish) [reply src/core/csv/reader.cc:794]: I don't think I quite understand that comment, can you try again? are you referring to the condition `*(ch+1) <= ' '`?

**Comment by pradkrish** [comment]
> One thing I don't like about the tab character `\t` is that it is both a whitespace character, and a legitimate separator. In fact, it is the second most popular separator in use after a comma.
> Thus, the definition of a "whitespace" changes when `\t` is a separator -- we don't want it to be treated as the separator in one part of a string, but as whitespace in another.
> 
> For example, the following is legitimate CSV fragment:
> 
> ```
> A\tB\tC\n
> \t2\t3\n
> 4\t5\t6\n
> ```
> 
> Note how `\t` at the beginning of a second line is not considered "whitespace at the start of a line".

I agree, neither the logic on this branch nor what's currently on master works for this kind of input. At the same time, what are chances such a fragment shows up "in the wild"? Perhaps rare, otherwise it would have shown up by now. One way to see would be that we postpone the full solution until it shows up.

File: src/core/csv/reader.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (st-pasha) [reply src/core/csv/reader.cc:794]: hmm, maybe I misread the logic in the conditional.

**Comment by pradkrish** [comment]
any more reviews left for this one? :thinking:

**Review by st-pasha — APPROVED** [review]

**Comment by st-pasha** [comment]
it's all good

**Pull request merged**